src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +429,424 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+  hrwr : qr.r.toNat < 2^wr
+  /-- The quotient is morally a `Bitvec wr`, and so has value less than `2^wr`. -/
+  hqwr : qr.q.toNat < 2^wr
+  /-- The low n bits of `n` obey the fundamental division equation. -/
+  hdiv : n.toNat >>> wn = d.toNat * qr.q.toNat + qr.r.toNat
+
+/-- A lawful DivModState implies `w > 0`. -/
+def DivModState.Lawful.hw {qr : DivModState w}
+    {h : DivModState.Lawful w wr wn n d qr} : 0 < w := by
+  have hd := h.hd
+  rcases w with rfl | w
+  · have hcontra : d = 0#0 := by apply Subsingleton.elim
+    rw [hcontra] at hd
+    simp at hd
+  · omega
+
+/-- An initial value with both `q, r = 0`. -/
+def DivModState.init (w : Nat) : DivModState w := {
+  q := 0#w
+  r := 0#w
+}
+
+/-- The initial state. -/
+def DivModState.Lawful.init (w : Nat) (n d : BitVec w) (hd : 0#w < d) :
+    DivModState.Lawful w 0 w n d (DivModState.init w) := {
+  hwrn := by omega,
+  hd := by assumption
+  hrd := by simp [BitVec.lt_def] at hd ⊢; assumption
+  hrwr := by simp [DivModState.init],
+  hqwr := by simp [DivModState.init],
+  hdiv := by
+    simp only [DivModState.init, toNat_ofNat, zero_mod, Nat.mul_zero, Nat.add_zero];
+    rw [Nat.shiftRight_eq_div_pow]
+    apply Nat.div_eq_of_lt n.isLt
+}
+
+/--
+A lawful DivModState with a fully consumed dividend (`wn = 0`) witneses that the
+quotient has been correctly computed.
+-/
+theorem DivModState.udiv_eq_of_lawful_zero {qr : DivModState w}
+    (h : DivModState.Lawful w w 0 n d qr) :
+    n.udiv d = qr.q := by
+  apply udiv_eq_of_mul_add_toNat h.hd h.hrd
+  have hdiv := h.hdiv
+  omega
+
+/--
+A lawful DivModState with a fully consumed dividend (`wn = 0`) witneses that the
+remainder has been correctly computed.
+-/
+theorem DivModState.umod_eq_of_lawful_zero {qr : DivModState w}
+    (h : DivModState.Lawful w w 0 n d qr) :
+    n.umod d = qr.r := by
+  apply umod_eq_of_mul_add_toNat h.hrd
+  have hdiv := h.hdiv
+  simp only [shiftRight_zero] at hdiv
+  exact hdiv.symm
+
+/-! ### LawfulShiftSubtract -/
+
+/-!
+A `LawfulShiftSubtract` is a `Lawful` DivModState that is also a legal input to the shift subtractor.
+So in particular, we must have at least one dividend bit left over `(0 < wn)`
+to perform a round of shift subtraction.
+-/
+
+/--
+The input to the shift subtractor is a legal input to `divrem`, and we also need to have an
+input bit to perform shift subtraction on, and thus we need `0 < wn`.
+-/
+structure DivModState.LawfulShiftSubtract (w wr wn : Nat) (n d : BitVec w) (qr : DivModState w)
+  extends DivModState.Lawful w wr wn n d qr : Type where
+  /-- Only perform a round of shift-subtract if we have dividend bits. -/
+  hwn_lt : 0 < wn
+
+/--
+In the shift subtract input, the dividend is at least one bit long (`wn > 0`), so
+the remainder has bits to be computed (`wr < w`).
+-/
+def DivModState.wr_lt_w {qr : DivModState w} (h : qr.LawfulShiftSubtract wr wn n d) : wr < w := by
+  have hwrn := h.hwrn
+  have hwn_lt := h.hwn_lt
+  omega
```

Comment:
```suggestion
  cases h <;> omega
```

Reply:
No :P That's a cool abuse of `cases` to expose the inhabitants, but, uh.